### PR TITLE
Upgrade Google Closure Compiler version.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Same as #2421, but using a custom backport of google closure compiler to Java 6.

**Comment by nicolasstucki** [comment]
Sources for the custom jar come from https://github.com/nicolasstucki/closure-compiler/commits/java-6

**Comment by nicolasstucki** [comment]
@sjrd comments in https://github.com/nicolasstucki/closure-compiler/commits/java-6 have been handeled.

**Comment by sjrd** [comment]
Apart from https://github.com/nicolasstucki/closure-compiler/commit/951b0f71ea87159f69e5f7ef286edb463ff40aea#commitcomment-18123308, the changes LGTM.

How should we proceed with this guy? What else do we need?

**Comment by nicolasstucki** [comment]
I believe that we just need to 
- [ ] change the `closure-compiler` project to be in `scala-js`
- [ ] publish this version of `closure-compiler`
- [ ] change this PR to use the published artefact

**Comment by sjrd** [comment]
After a very long dive into Maven/OSSRH documentation and configuration, I finally managed to stage the published artifact. Could you test with this resolver:
```scala
resolvers += "Scala.js GCC staging" at "https://oss.sonatype.org/content/repositories/orgscala-js-1054/"
```
and a dependency on
```scala
"org.scala-js" % "closure-compiler-java-6" % "v20160517"
```
?

**Comment by sjrd** [comment]
Seems to be working https://scala-webapps.epfl.ch/jenkins/job/scalajs-task-worker/131220/consoleFull :)

**Comment by nicolasstucki** [comment]
@sjrd last thing now, you need to release the `closure-compiler-java-6` and I need to remove the staging resolver.

**Comment by sjrd** [comment]
checksizes issues, probably due to the rebase on top of #2488.

**Comment by nicolasstucki** [comment]
Yes, it was because of that rebase.

**Comment by sjrd** [comment]
LGTM

**Comment by sjrd** [comment]
I suggest we merge #2468 first.

File: project/Build.scala [hunk 1]
```diff
@@ -1496,13 +1496,12 @@ object Build {
 
             trgDir
           },
-
```
Comment (sjrd) [inline project/Build.scala]: Just noticed the useless diff here. This empty line should be kept.

**Comment by sjrd** [comment]
This should now be rebased (and you can restore that blank line in the process).

**Comment by sjrd** [comment]
Nightly: https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/4295/

File: tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureAstTransformer.scala [hunk 2]
```diff
@@ -264,7 +264,7 @@ private[closure] class ClosureAstTransformer(relativizeBaseURI: Option[URI]) {
     @inline def ctorDoc(node: Node) = {
```
Comment (gzm0) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureAstTransformer.scala]: Can we remove the argument if we don't need it anymore?

**Comment by nicolasstucki** [comment]
The following tests failed in the nightly
```
setJavaVersion 1.7
sbt ++2.11.8 package "partestSuite/testOnly -- --fullOpt --showDiff"
!!  47 - run/bugs.scala                            [output differs]
!!  66 - run/contrib674.scala                      [output differs]
!!  98 - run/finally.scala                         [output differs]
!! 133 - run/lazy-locals.scala                     [output differs]
!! 274 - run/patmatnew.scala                       [output differs]
!! 281 - run/run-bug4840.scala                     [output differs]
!! 282 - run/range.scala                           [output differs]
!! 289 - run/spec-nlreturn.scala                   [output differs]
!! 386 - run/t2333.scala                           [output differs]
!! 585 - run/t5380.scala                           [output differs]
!! 723 - run/t6863.scala                           [output differs]
!! 808 - run/t8245.scala                           [output differs]
!! 829 - run/t8845.scala                           [output differs]
!! 867 - run/try-2.scala                           [output differs]
!! 869 - run/try.scala                             [output differs]
!! 875 - run/tuple-zipped.scala                    [output differs]
!! 880 - run/unreachable.scala                     [output differs]
```

```
setJavaVersion 1.8
sbt ++2.12.0-M5 package "partestSuite/testOnly -- --fullOpt --showDiff"
!!  44 - run/bugs.scala                            [output differs]
!!  66 - run/contrib674.scala                      [output differs]
!!  97 - run/finally.scala                         [output differs]
!! 132 - run/lazy-locals.scala                     [output differs]
!! 273 - run/patmatnew.scala                       [output differs]
!! 281 - run/range.scala                           [output differs]
!! 297 - run/spec-nlreturn.scala                   [output differs]
!! 392 - run/t2333.scala                           [output differs]
!! 432 - run/t3150.scala                           [output differs]
!! 518 - run/t4148.scala                           [output differs]
!! 591 - run/t5380.scala                           [output differs]
!! 727 - run/t6863.scala                           [output differs]
!! 813 - run/t8245.scala                           [output differs]
!! 836 - run/t8845.scala                           [output differs]
!! 885 - run/try-2.scala                           [output differs]
!! 887 - run/try.scala                             [output differs]
!! 900 - run/unreachable.scala                     [output differs]
```

File: tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureAstTransformer.scala (hunk 2 above)
Comment (nicolasstucki) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureAstTransformer.scala]: Yes, we can.

**Comment by nicolasstucki** [comment]
The source of the issue is that closure is emitting warnings thats get caught in the partest output.
As far as I can see all of these are `Closure: JSC_UNREACHABLE_CODE. unreachable code at ...`.
Those warings are generated in [CheckUnreachableCode.java](https://github.com/scala-js/closure-compiler/blob/java-6/src/com/google/javascript/jscomp/CheckUnreachableCode.java#L63).

**Comment by gzm0** [comment]
Can we configure it to remove unreachable code?

**Comment by nicolasstucki** [comment]
That is what I'm exploring right now.

**Comment by sjrd** [comment]
I believe it *does* remove unreachable code. But it warns about it for the developer to go clean up their source codebase. In our case, we should instead configure it to not warn. We already have something like that here: https://github.com/scala-js/scala-js/blob/608baf718e4366dfdcb0745f13facb2b438344f6/tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/LoggerErrorManager.scala#L16

**Comment by sjrd** [comment]
Oh and a nicer configuration here: https://github.com/scala-js/scala-js/blob/608baf718e4366dfdcb0745f13facb2b438344f6/tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureLinkerBackend.scala#L137

**Comment by gzm0** [comment]
Yeah, the ErrorManager :) If you are touching this thing anyways, you might want to have a look whether GCC sanitized their interface and made some of the error types public. I think there is a way to disable reports upfront.

**Comment by gzm0** [comment]
Ah, you found it :) JSC_FRACTIONAL_BITWISE_OPERAND didn't have this config, I recall. But maybe that has changed as well.

**Comment by nicolasstucki** [comment]
Yes, it seems to be woring by it to `LoggerErrorManager.report`

File: tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/LoggerErrorManager.scala [hunk 3]
```diff
@@ -13,10 +13,12 @@ private[closure] class LoggerErrorManager(private val log: Logger)
     * package private.
     */
   override def report(level: CheckLevel, error: JSError): Unit = {
-    if (error.getType.key == "JSC_FRACTIONAL_BITWISE_OPERAND")
+    if (error.getType.key == "JSC_FRACTIONAL_BITWISE_OPERAND" ||
+        error.getType.key == "JSC_UNREACHABLE_CODE") {
```
Comment (sjrd) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/LoggerErrorManager.scala]: I guess there was no method of `options` for either of those, like we do [here](https://github.com/scala-js/scala-js/blob/608baf718e4366dfdcb0745f13facb2b438344f6/tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureLinkerBackend.scala#L137)?

Comment (gzm0) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/LoggerErrorManager.scala]: Please update the comment above.

Comment (nicolasstucki) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/LoggerErrorManager.scala]: There was no method. That was my first approach.

**Comment by sjrd** [comment]
LGTM, assuming you have locally tested with `--fullOpt` the partests that failed before.

**Comment by nicolasstucki** [comment]
I did it once, but did some minor changes afterwards. I will retest them to make sure.

**Comment by nicolasstucki** [comment]
All the listed partest passed.

**Comment by sjrd** [comment]
Weekly, just to be sure: https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/4299/

**Comment by nicolasstucki** [comment]
The weekly failed epicly on `2.11.0`, `2.11.1` and `2.11.2`. Neg tests fail compilation but also fail the partest. 
 
@sjrd I wonder if that was not introduced when changing the `partest` version to upgrade to `2.12.-M5`.

```
# starting 842 tests in neg
!!   1 - neg/abstract-class-error                  [compilation failed]
!!   2 - neg/abstract-class-2.scala                [compilation failed]
!!   3 - neg/abstract-concrete-methods.scala       [compilation failed]
!!   4 - neg/abstract-vars.scala                   [compilation failed]
!!   5 - neg/abstract-report.scala                 [compilation failed]
!!   6 - neg/abstract-explaintypes.scala           [compilation failed]
!!   7 - neg/abstract-report2.scala                [compilation failed]
!!   8 - neg/accesses2.scala                       [compilation failed]
!!   9 - neg/abstract-inaccessible.scala           [compilation failed]
!!  10 - neg/ambiguous-float-dots2.scala           [compilation failed]
...
```

**Comment by sjrd** [comment]
It seems it failed just as epically on master on Saturday. So yes that is highly probable. I guess that means we need to fix it on master first.

**Comment by sjrd** [comment]
This should be rebased to get the fix to weekly.

**Comment by sjrd** [comment]
Weekly: https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/4315/

**Comment by sjrd** [comment]
Weekly basically passed. There is one failed config, but the failures are disjoint between its two runs, so they must be spurious.

**Comment by siriux** [comment]
Is there any significant difference on the benchmarks in speed or size using the new GCC version?

**Comment by sjrd** [comment]
@siriux Actually, in some benchmarks, yes:

Benchmark | fastOpt | fullOpt old GCC | fullOpt new GCC | JVM
------------ | ------------- | --- | --- | ---
Richards | 106 | 133 | 134 | 62
DeltaBlue | 269 | 420 | 217 | 147
Tracer | 1063 | 964 | 820 | 608
Sudoku | 6185 | 5902 | 5428 | 1112
SHA-512 | 91863 | 81600 | 79846 | 8873

* The most important difference can be seen on DeltaBlue, with a ~2x speedup, which makes it faster than fastOpt instead of significantly slower, so that's good.
* Richards did not move. The fullOpt version is still slower than the fastOpt version.
* Tracer receives a significant speedup of 15%, and gets very close to the JVM version, at 1.35x.
* Sudoku receives a 10% speedup.
* SHA-512 does not significantly move.

**Comment by siriux** [comment]
That's great ! Do you have the sizes too?

Any know reason for Sudoku and specially SHA-512 to be much slower than JVM? The ~2x of the other benchmarks seems normal, but 5x or 10x is pretty high. 

Anyway, that's probably for other issue, as it was already the case with the old version of GCC and fastOpt.

**Comment by sjrd** [comment]
You can see the changes in sizes here: https://github.com/scala-js/scala-js/pull/2490/files#diff-e00c3bd6d919188ac2349d6e736145ae
Basically, not much movement, here.

> Any know reason for Sudoku and specially SHA-512 to be much slower than JVM?

SHA-512, yes, definitely: it's the Longs. We've made *tremendous* improvements to the Longs in #2488, but they're still implemented in user-space, whereas JVM Longs are implemented with the processor's 64-bit instructions. Being only 10x slower is actually a world-class achievement.

For Sudoku, no idea. I never really investigated that one.

**Pull request merged**